docs/README.md

```diff
@@ -0,0 +1,38 @@
+# Nomad Codebase Documentation
+
+This directory contains some documentation about the Nomad codebase,
+aimed at readers who are interested in making code contributions.
+
+If you're looking for information on _using_ using, please instead refer
+to [the main Nomad website](https://nomadproject.io).
+
+## Architecture
+
+The code for Nomad's major components is organized as:
+
+* `api/` provides a Go client for Nomad's HTTP API.
+* `client/` contains Nomad's client agent code.
+* `command/` contains Nomad's CLI code.
+* `nomad/` contains Nomad's server agent code.
+
+The high level control flow for many Nomad actions (via the CLI or UI) are:
+
+```
+# Read actions:
+Client -> HTTP API -> RPC -> StateStore
+
+# Actions that change state:
+Client -> HTTP API -> RPC -> Raft -> FSM -> StateStore
```

Comment:
Might be worth pointing out that that internal non-user-http-triggered changes go through RPC without HTTP API.

Reply:
I couldn't really figure out how to succinctly mention it so skipping for now. This didn't seem very helpful:

```
# Internode communication
RPC <-> RPC
```